pip_run/commands.py

```diff
@@ -15,7 +16,44 @@ def _is_python_arg(item: str):
     to Python and not to pip install.
     """
     path = pathlib.Path(item)
-    return path.is_file() and path.suffix == '.py'
+    # start with the tests which do very minimal IO
+    if not path.exists():
+        return False
+    if path.suffix == ".py":
+        return True
+
+    # now, sniff the first line of the file and return true if it looks like a
+    # python script based on the shebang line
+    return _has_python_shebang(path)
+
+
+def _has_python_shebang(path: pathlib.Path) -> bool:
+    try:
+        with path.open('rb') as fp:
+            first_line_bytes = fp.readline()
+        first_line = first_line_bytes.decode('utf-8')
+    except UnicodeDecodeError:
+        return False
+    if not first_line.startswith('#!'):
+        return False
+
+    cmd = shlex.split(first_line[2:])
+    if cmd[0] == "/usr/bin/env":
+        cmd = cmd[1:]
+        if cmd[0] == "-S":
+            cmd = cmd[1:]
+        command_name = cmd[0]
+    else:
+        command_name = cmd[0].split('/')[-1]
+
+    return command_name in (
+        "python",
+        "pip-run",
```

Comment:
I'm especially reluctant to include pip-run in the check. That name makes sense in this context, but I also have hopes of porting this to `pip run` (a subcommand of pip). This is yet another reason to consider reducing the specificity of this check.

It _could_ be sufficient to only check for `py`, but that would require that users use something like "#!/usr/bin/env py -m pip-run" to cause "py" to appear in the shebang.

I'm leaning toward let's just not be validating the shebang.

Reply:
I understand being hesitant to include `pip-run` if `pip run` is still your goal. One of my goals is to be able to have shebangs of the form `#!/usr/bin/env pip-run` -- I think that's the most intuitive way to try to make a "self launching pip-run script" -- so I don't like the idea of requiring a specific usage of the `py` launcher or `python -m pip-run`. 

Cross-checking with the comment above, I think this boils down to just checking for `#!` but without validating the contents of the shebang line.

Let me know if you'd prefer just checking `#!` or `#!` with a fallback of `ast.parse`. That's the direction I'm leaning right now, but of course I'm open to being persuaded of a different approach.

Reply:
Let's go with simply checking for shebang and we can consider adding a parse check later if needed.

Reply:
I just got back around to this.

In addition to reducing the check to just a shebang line, I added support for shebangs with leading BOM bytes.
It maybe isn't strictly necessary, but as I was removing a lot of the other handling I realized I hadn't even thought about it in the initial implementation.

Let me know if you'd prefer to drop the BOM checking for simplicity.

Reply:
As mentioned in a subsequent review, I've decided to go with a less intrusive support for UTF-8 BOM only.